File: python_pipeline/run_all.py
#!/usr/bin/env python3
# =============================================================================
# PYTHON PIPELINE - ORCHESTRATOR (equivalent of run_all.R)
#
# DEFAULT (current project focus): AR benchmark + updated standard R-MIDAS
# (Almon over the AR lag set J = {1,2,3,7}), then the summary table and the
# combination / subperiod / GR evaluation restricted to the r_midas family.
#
# The other stages (r_midas_ext, ru_midas, surveys_h) remain registered and
# can be selected explicitly, e.g.:
#   STAGES=ar,r_midas,ru_midas,summary python3 run_all.py
#
# Env knobs (forwarded to every stage): COUNTRY_NAME, SPEC_LIMIT, N_CORES,
# WINDOW_DAYS, RMIDAS_DISCOUNT, ORIGIN_LIMIT, FAMILIES.
# =============================================================================

import importlib
import os
import time

STAGE_MODULES = {
    "ar": "ar_benchmark",
    "r_midas": "r_midas",
    "r_midas_ext": "r_midas_extended",
    "ru_midas": "ru_midas",
    "surveys_h": "r_midas_surveys_h",
    "summary": "summary_table",
    "combination": "combination_subperiod_gr",
}


def _load_stage(stage):
    """Import only the requested stage, so optional files may be absent."""
    module = importlib.import_module(STAGE_MODULES[stage])
    return module.main


# Current focus: AR + updated R-MIDAS only (other stages via STAGES env var).
DEFAULT_ORDER = ["ar",
                 "r_midas",
                 "summary",
                 "combination"]


def main():
    stages = os.environ.get("STAGES", "")
    order = ([s.strip() for s in stages.split(",") if s.strip()]
             if stages else DEFAULT_ORDER)
    for st in order:
        if st not in STAGE_MODULES:
            print(f"Unknown stage '{st}' (known: {list(STAGE_MODULES)})")
            continue
        print(f"\n{'='*70}\nSTAGE: {st}\n{'='*70}")
        # Restrict the combination stage to the r_midas family unless the
        # user set FAMILIES explicitly.
        old_families = os.environ.get("FAMILIES")
        set_families_here = st == "combination" and old_families is None
        if set_families_here:
            os.environ["FAMILIES"] = "r_midas"
        t0 = time.time()
        try:
            _load_stage(st)()
        finally:
            if set_families_here:
                os.environ.pop("FAMILIES", None)
        print(f"stage {st} done in {time.time() - t0:.1f}s")
    print("\nAll requested stages done.")


if __name__ == "__main__":
    main()

File: r_pipeline/midas_codes/run_all.R
#!/usr/bin/env Rscript
# =============================================================================
# v2 PIPELINE — ORCHESTRATOR (stage-selectable)
#
# DEFAULT (current project focus): AR benchmark + updated standard R-MIDAS
# only, followed by the AR+R-MIDAS summary and the combination / subperiod /
# Giacomini-Rossi evaluation restricted to the r_midas family.
#
#   ar           ar_benchmark.R      AR(1,2,3,7) + calendar; ar_dum = benchmark
#   r_midas      r_midas.R           STANDARD R-MIDAS — exp-Almon on the AR lag
#                                    set J = {1,2,3,7} of y (the j values enter
#                                    the Almon exponent, see model.tex);
#                                    gas_lag1 / brent_lag1 linear; LF linear;
#                                    18 calendar dummies.  Family: "R_MIDAS".
#   summary      summary_table_ar_r_midas.R   AR + R_MIDAS metrics only
#   combination  combination_subperiod_gr.R   combos + subperiods + GR test
#                                    (FAMILIES defaults to "r_midas" here)
#
# Other stages, NOT run by default but selectable via the STAGES env var:
#   ru_midas     ru_midas.R            RU-MIDAS position dummies
#   r_midas_ext  r_midas_extended.R    extended R-MIDAS (Almon on y AND HF)
#   summary_all  summary_table.R       summary across all families
#
# Usage:
#   Rscript run_all.R                                    # default subset
#   STAGES="ar,r_midas,ru_midas,summary_all" Rscript run_all.R
#
# Knobs: COUNTRY_NAME ("Germany" | "Italy" | "ALL"), N_CORES, SPEC_LIMIT,
# WINDOW_DAYS, RMIDAS_DISCOUNT — see common_utils.R.
#
# Runtime estimate (one country, N_CORES = 9, 256 specs):
#   ar          : < 1 min
#   r_midas     : with the FWL-accelerated NLS expect tens of minutes rather
#                 than the 8-35 h of the pre-acceleration runs
#   summary     : < 30 sec
#   combination : ~1-2 min
# =============================================================================

args <- commandArgs(trailingOnly = FALSE)
farg <- args[grep("^--file=", args)]
v2.boot.path <- if (length(farg) > 0) {
  raw <- gsub("~+~", " ", sub("^--file=", "", farg[1]), fixed = TRUE)
  file.path(dirname(suppressWarnings(normalizePath(raw, winslash = "/", mustWork = FALSE))),
            "v2_paths.R")
} else file.path(getwd(), "v2_paths.R")
source(v2.boot.path)

STAGE_FILES <- c(
  ar          = "ar_benchmark.R",
  r_midas     = "r_midas.R",
  ru_midas    = "ru_midas.R",
  r_midas_ext = "r_midas_extended.R",
  summary     = "summary_table_ar_r_midas.R",
  summary_all = "summary_table.R",
  combination = "combination_subperiod_gr.R"
)
DEFAULT_ORDER <- c("ar", "r_midas", "summary", "combination")

stages_env <- Sys.getenv("STAGES", unset = "")
order <- if (nzchar(stages_env)) {
  trimws(strsplit(stages_env, ",")[[1]])
} else DEFAULT_ORDER

t0 <- Sys.time()
cat("=== v2 pipeline start: ", format(t0), " ===\n")
cat("stages: ", paste(order, collapse = " -> "), "\n")

for (st in order) {
  if (!st %in% names(STAGE_FILES)) {
    cat(sprintf("\nUnknown stage '%s' (known: %s) — skipped.\n",
                st, paste(names(STAGE_FILES), collapse = ", ")))
    next
  }
  fp <- file.path(script_dir, STAGE_FILES[[st]])
  cat(sprintf("\n[%s] %s\n", st, STAGE_FILES[[st]]))
  if (!file.exists(fp)) {
    cat("  file not found — stage skipped.\n")
    next
  }
  # Restrict the combination stage to the r_midas family unless the user
  # already set FAMILIES explicitly.
  old_families <- Sys.getenv("FAMILIES", unset = NA_character_)
  set_families_here <- st == "combination" && is.na(old_families)
  if (set_families_here) Sys.setenv(FAMILIES = "r_midas")
  ts <- Sys.time()
  tryCatch(
    source(fp),
    finally = if (set_families_here) Sys.unsetenv("FAMILIES")
  )
  cat(sprintf("  stage '%s' done in %s\n", st,
              format(round(difftime(Sys.time(), ts, units = "mins"), 1))))
}

t1 <- Sys.time()
cat(sprintf("\n=== v2 pipeline done in %s ===\n",
            format(round(difftime(t1, t0, units = "mins"), 1))))
